src/__test__/index.test.ts

```diff
@@ -295,24 +238,116 @@ describe('test rpc requests to Snap', () => {
+			const initPairingReq = request({
+				method: InternalMethod.TssInitPairing,
+				origin: ORIGIN,
+				params: [{ isRePair: false }],
+			});
 
-			let legacySignResult: any = await keyring.signTransaction(
-				mockLegacyTx,
-				runTssSign,
+			const ui = await initPairingReq.getInterface();
+			expect(ui.type).toBe(DialogType.Confirmation);
+			const prompt = INIT_PAIR_PANEL_HEADING;
+			const description = INIT_PAIR_PANEL_DESCRIPTION;
+			expect(ui).toRender(
+				panel([
+					heading(prompt),
+					divider(),
+					...description.map((t) => text(t)),
+				]),
 			);
+
+			await ui.ok();
+
+			const initPairingJson: any = (await initPairingReq).response;
+			const initPairingResult =
+				initPairingJson.result as InitPairingResponse;
+			const qrCode = initPairingResult.qrCode;
+			const qrCodeObj = JSON.parse(qrCode) as QrCode;
+
+			expect(qrCodeObj.pairingId).toEqual(expect.any(String));
+			expect(qrCodeObj.webEncPublicKey).toEqual(expect.any(String));
+			expect(qrCodeObj.signPublicKey).toEqual(expect.any(String));
+
+			// Test run pairing
+			await simulator.signIn();
+			await simulator.pairing(qrCode);
+
+			const runPairingReq = request({
+				method: InternalMethod.TssRunPairing,
+				origin: ORIGIN,
+			});
+
+			const runPairingJson: any = (await runPairingReq).response;
+			const runPairingResult =
+				runPairingJson.result as RunPairingResponse;
+			expect(runPairingResult.deviceName).toEqual(DEVICE_NAME);
+			expect(runPairingResult.address).toBeNull();
+
+			// Test key generation
+			const keygenReq = request({
+				method: InternalMethod.TssRunKeygen,
+				origin: ORIGIN,
+			});
+			await simulator.keygen();
+
+			const keyGenJson: any = (await keygenReq).response;
+			const runKeyGenResult = keyGenJson.result as RunKeygenResponse;
+			expect(runKeyGenResult.address).toEqual(expect.any(String));
+
+			request({
+				method: InternalMethod.TssRunBackup,
+				origin: ORIGIN,
+			});
+
+			// Test signing
+			const keyshareReq = request({
+				method: InternalMethod.E2eTestGetKeyShare,
+				origin: ORIGIN,
+			});
+			const keyshareReqJson: any = (await keyshareReq).response;
+			const keyshareResult = keyshareReqJson.result as {
+				distributedKey: DistributedKey;
+				pairingData: PairingData;
+			};
+
+			const accountId = keyshareResult.distributedKey.accountId;
+			const keyring = new SimpleKeyring({
+				wallets: {
+					[accountId]: {
+						...MOCK_WALLET_ACCOUNT,
+						distributedKey: keyshareResult.distributedKey,
+					},
+				},
+				requests: {},
+			});
+
+			const pairingData = keyshareResult.pairingData;
+			const runTssSign = genMockRunTssSign(pairingData);
+
+			// Legacy sign
+
+			let legacySignResult: any = null;
+			keyring
+				.signTransaction(mockLegacyTx, runTssSign)
+				.then((resp: any) => {
+					legacySignResult = resp;
+				})
+				.catch((err) => {
+					console.log('err', err);
+				});
+			simulator.sign();
+
+			while (!legacySignResult) {
+				await delay(0);
```

Comment:
What is the use of this?

Reply:
The signing happens by exchanging messages between 2 parties. Then, we need to wait for those communication rounds to be finished to get the result without moving to the next case

Reply:
Why do I use "while" loop here but not 2 async/await call: 
Because we're running in the same thread and try to simulate 2 actor communication (SimpleKeyring instance and app simulator).
If we use async/await, JS can't not delegate the async works of these 2 actors like we're running in 2 different threads, instead, they will wait for each other => we will fail in exchanging messages.

Reply:
Please add a comment why this is done in the code.